utils/v2/connector/impl/RedStonePriceFeed.ts

```diff
@@ -65,11 +66,45 @@ export class RedStonePriceFeed implements PriceFeedConnector {
     // We need to get signer public key firstly
     // It will be used to pre-verify signatures off-chain
     await this.getSigner();
-    
-    // Fetching data from all sources simultaneously with timeout
+
     const timeoutMilliseconds =
       this.priceFeedOptions.dataSources?.timeoutMilliseconds
       || DEFAULT_TIMEOUT_MILLISECONDS;
+    const useFirstValid =
+      this.priceFeedOptions.dataSources!.valueSelectionAlgorithm === "first-valid";
+
+    const selectedResponse = useFirstValid
+      ? await this.fetchFirstValid(timeoutMilliseconds)
+      : await this.fetchAllAndSelectValid(timeoutMilliseconds);
+
+    return convertResponseToPricePackage(selectedResponse, this.cachedSigner!);
+  }
+
+  private async fetchFirstValid(timeoutMilliseconds: number): Promise<SignedDataPackageResponse> {
+    let fetcherIndex = 0;
+    const promises = this.fetchers.map(fetcher => {
+      fetcherIndex++;
+      return (async () => {
+        const response = await fetcher.getLatestDataWithTimeout(
+          this.providerId,
+          timeoutMilliseconds);
+        const isValid = validateDataPackage(response, this.priceFeedOptions, this.cachedSigner!);
+        if (isValid) {
+          return response;
+        } else {
+          console.warn("Invalid response: " + JSON.stringify(response));
+          throw new Error(
```

Comment:
Shouldn't we pass fetcherIndex as an argument to the async() method?

Reply:
I am not sure I understand the question. Can you please provide more details?

Reply:
The idea was to check if we shouldn't wrap the argument and connect it to promise's context:
https://stackoverflow.com/questions/35318442/how-to-pass-parameter-to-a-promise-function

to be sure that changing promise will read the fetcherIndex with value at the time it's passed to the promise.

Reply:
Oh, right sorry, just fixed. Thank you for noticing!